### PR TITLE
Port codebase to tui

### PR DESCRIPTION
**Pull request by Plecra**
Windows support is the primary motivation here, but the `tui` crate is also used much more widely in the ecosystem and provides some handy abstractions.

I've kept changes in this PR to a minimum, replacing `tuikit` calls with their closest `tui` equivalents. There's probably more that could be done to make the code more idiomatic `tui` code.

### CODE REVIEW

**Comment by Plecra** [comment]
Here's a demo gif for anyone interested 
7928a711-9bff-4f91-a977-d64f580e0db7

File: src/editor/keystroke_log.rs [hunk 1]
```diff
@@ -71,27 +78,10 @@ pub struct KeyStrokeLog {
...
-
-    /// Sets and enforces the max entry limit
-    pub fn set_max_entries(&mut self, max_entries: usize) {
-        self.max_entries = max_entries;
-        self.enforce_entry_limit();
-    }
-
-    /// Draw a log of recent keystrokes to a given terminal at a given location
-    pub fn render(&self, term: &Term, row: usize, col: usize) {
+impl tui::widgets::Widget for &'_ KeyStrokeLog {
```
Comment (kneasle) [inline src/editor/keystroke_log.rs:83]: Tiny nitpick: a blank line between the struct and `impl` would be nice - I'm not sure why rustfmt doesn't enforce this by default, but hey.

File: src/editor/keystroke_log.rs [hunk 2]
```diff
@@ -112,34 +102,58 @@ impl KeyStrokeLog {
...
                 &e.description,
-                Attr::default().fg(e.color),
-            )
-            .unwrap();
+                Style::default().fg(e.color),
+            );
         }
     }
+}
+impl KeyStrokeLog {
```
Comment (kneasle) [inline src/editor/keystroke_log.rs:142]: Also a newline here would be nice :grin:.

File: src/editor/mod.rs [hunk 3]
```diff
@@ -282,7 +290,13 @@ impl<'arena, Node: Ast<'arena> + 'arena> Editor<'arena, Node> {
...
-        self.term.show_cursor(true).unwrap();
-        self.term.present().unwrap();
+        self.term.as_mut().unwrap().show_cursor().unwrap();
+        crossterm::execute!(
+            self.term.as_mut().unwrap().backend_mut(),
+            terminal::LeaveAlternateScreen
+        )
+        .unwrap();
+        self.term.as_mut().unwrap().flush().unwrap();
+        terminal::disable_raw_mode().unwrap();
```
Comment (kneasle) [inline src/editor/mod.rs:300]: :laughing: This is a truly ungodly amount of unwrapping but I guess if the terminal bugs out somehow then Sapling is screwed regardless and it doesn't matter.

File: src/main.rs [hunk 4]
```diff
@@ -28,9 +28,9 @@ use serde_json::Value;
 /// [`Editor::run`].
 fn main() {
     // Initialise the logging and startup
-    pretty_env_logger::formatted_builder()
-        .filter_level(log::LevelFilter::Debug)
-        .init();
+    // pretty_env_logger::formatted_builder()
+    //     .filter_level(log::LevelFilter::Debug)
+    //     .init();
```
Comment (kneasle) [inline src/main.rs:33]: I feel like we should probably get the logging back... but this brings me up a point.  Is there a better way to handle logging when running a TUI application? I guess writing to a log file would be the best option, but I'm not experienced enough to know how that would work.

**Review by kneasle — APPROVED** [review]
Very nice!  Just a few minor formatting tweaks and I'd be happy to merge.

**Comment by kneasle** [comment]
Actually thinking about this, I could probably fix the formatting myself and merge... but the question about the logging is relevant so I'll probably leave this open for a bit.

**Comment by Plecra** [comment]
I focused on changing the layout of the code as little as possible in this PR, but restructuring to avoid the unwrapping should be very doable (and worth doing!)

I actually have a plan for logging in mind. We can collect the `log` messages and display them in a new 'system logs' pane. This'll be much easier to do when we've fully integrated with `tui`'s layout logic. Imo this should be done in another PR. If logging is needed in the meantime, there are crates that implement the logger as writing to a file.

**Comment by kneasle** [comment]
This looks good!  I'll merge this now - you're welcome to do any more refactoring if you want; all refactoring is fair game IMO provided that it makes the code better (which refactoring nearly always does).

**Pull request merged**